### PR TITLE
feat(RELEASE-971): add task to push disk images

### PR DESCRIPTION
**Pull request by johnbieren**
This commit adds a task to push disk images via an InternalRequest. The steps done in the pull-disk-images task will be removed in a future commit as they are now done in the InternalRequest this task calls.

### CODE REVIEW

**Comment by johnbieren** [comment]
Opening this early before the tests are written so I can get early feedback as this relates to the ai effort @konflux-ci/release-service-maintainers FYI

File: tasks/push-disk-images/push-disk-images.yaml
Comment (scoheb) [inline tasks/push-disk-images/push-disk-images.yaml:52]: not stage?

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [inline tasks/push-disk-images/push-disk-images.yaml:1]: Other than what Scott asked, it looks straightforward to me. But maybe a dumb question: Why do we need internal requests for this? I guess the limitation is that it needs to run on red hat network - to have access to pulp (and maybe something else too). But couldn't we just require users to be on the internal Konflux cluster (and enable access to the required internal services in the fw settings)?

File: tasks/push-disk-images/push-disk-images.yaml
Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:52]: Per https://issues.redhat.com/browse/RHELDST-24850, the exodus gw is the same for both prod and stage, we just use different pulp and env values

File: tasks/push-disk-images/push-disk-images.yaml
Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:1]: Yes, it is because pulp is internal only. I'm not sure about the last question in your post. Since when do we have the ability to have users run release-service-catalog pipelines *only* on internal clusters?

File: tasks/push-disk-images/push-disk-images.yaml
Comment (scoheb) [reply tasks/push-disk-images/push-disk-images.yaml:52]: maybe add a comment so ppl don't ask same question at a later time?

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:1]: I don't think we have the ability to limit that, no. But even today, as a user you need to know if you want to use the public or internal cluster. I remember seeing a thread in the konflux-users channel when somebody was asking this and somebody else told him it depends on if you want this or that (but I don't remember the specifics). For example, if you're on the private cluster, you will have to use the prod-internal server for pyxis.

Oh yes, I think the main use case is if you want to access the internal gitlab (have the code there), then you need to use the internal cluster. So perhaps this could be similar?

Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:1]: I would assume we can still create InternalRequests from the internal cluster and the internal request pipelines are guaranteed to run on the internal cluster. So I guess I am not following the ask here - do not use an internal request, have users do this magic we don't currently know to force them to run on prod-internal or whatever we call it, and then just have it do the same logic that the internal request does? I don't see the benefit of that and it locks out those on the public cluster

Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:1]: The benefit is that it's much easier to deal with everything in the catalog repo. Not only for development, but also for debugging - it's definitely harder to debug when something fails inside of an internal request. Also, the internal tekton catalog repo that we have is still not on par with the main one - e.g. no task tests. 

Of course the counter argument is that this is what we've been doing for services inside red hat network - e.g. IIB. So it makes sense to do the same here.

But ever since I heard of having an internal konflux cluster I thought that maybe eventually we might want to get rid of internal requests all together - everything would be easier that way.

Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:1]: good points, but I think it is better for a standup meeting and if agreed on, moving everything. I don't think we should have the conversation here (especially since this is rhel-ai which is very high priority)

Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:1]: Yeah, I agree. And I definitely don't want to hold this over this. So yeah, we should probably have a more general discussion about this.

File: tasks/push-disk-images/push-disk-images.yaml
Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:52]: added

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [inline tasks/push-disk-images/push-disk-images.yaml:47]: I was looking at the corresponding PR here: https://github.com/hacbs-release/app-interface-deployments/pull/140

And I was wondering why exodusGwEnv is a parameter while pulp url is stored in the secret. If the pulp secret can contain the url, then maybe the exodus secret could contain the env and you wouldn't need a special parameter? I'm not saying it's better, just a thought.

Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:47]: The reason why I took this approach is because right now, we only need two exodus secrets, prod and nonprod. For staging, we use the prod secrets (creds + url) but with exodusGwEnv set to pre. So I could make a prod exodus secret and a stage one and include env in there, but I thought `live` and `pre` are so *not* secret that it would be easiest to do it this way as other than those values, the secrets would be identical

Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:47]: OK, makes sense.

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:83]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:95]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:105]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:111]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:123]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:83]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:95]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:105]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:111]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:117]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-stage.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-stage.yaml:123]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-qa.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-qa.yaml:117]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:123]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:117]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:111]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:105]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:95]: fixed

File: tasks/push-disk-images/tests/test-push-disk-images-production.yaml
Comment (johnbieren) [reply tasks/push-disk-images/tests/test-push-disk-images-production.yaml:83]: fixed

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [inline tasks/push-disk-images/push-disk-images.yaml]: I would suggest dropping the `-e`. The conditions below will make sure that a nice error message is printed if this is not set or set incorrectly. I think that's a better user experience than just failing here without any explanation if it's unset.

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [inline tasks/push-disk-images/push-disk-images.yaml:86]: The internal-request command will fail if the IR fails, so there's no point in checking it here again - you will never get here if it fails.
See here: https://github.com/konflux-ci/release-service-utils/blob/main/utils/wait-for-ir#L126
https://github.com/konflux-ci/release-service-utils/blob/main/utils/wait-for-ir#L135

File: tasks/push-disk-images/tests/mocks.sh
Comment (mmalina) [inline tasks/push-disk-images/tests/mocks.sh:22]: Depending on how you'll respond to my comment about the status check being redundant, you may want to use an approach similar to what we have for rh-sign-image.

File: tasks/push-disk-images/push-disk-images.yaml
Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml:86]: https://github.com/hacbs-release/app-interface-deployments/blob/main/internal-services/catalog/pulp-push-disk-images-task.yaml#L107 the IR will never fail. A failure here is characterized by a successful IR with failure in its results

File: tasks/push-disk-images/push-disk-images.yaml
Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml]: Good point, will update

Comment (johnbieren) [reply tasks/push-disk-images/push-disk-images.yaml]: fixed

File: tasks/push-disk-images/push-disk-images.yaml
Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:86]: ah, I just realized this now looking at another PR... my bad.

Comment (mmalina) [reply tasks/push-disk-images/push-disk-images.yaml:86]: I mean there is always a chance the IR fails - not likely, but it could be some tekton error. So we could have a test for that as well. But it's more useful to test the scenario you're testing here.

File: tasks/push-disk-images/tests/mocks.sh
Comment (mmalina) [reply tasks/push-disk-images/tests/mocks.sh:22]: Closing this based on the conversation in https://github.com/konflux-ci/release-service-catalog/pull/471#discussion_r1689328270

**Review by mmalina — APPROVED** [review]

**Review by happybhati — APPROVED** [review]

**Review by happybhati — APPROVED** [review]
/test release-pipelines-e2e-suite

**Pull request merged**
